"Building multiple Progressive Web Apps on the same domain" article.

demianrenzulli opened the pull request:
Fixes #5164

@jpmedley article is ready for review. We are still waiting for the assets of the other two authors before merging.

cc // @petele @kaycebasques

jpmedley commented on src/site/content/en/blog/building-multiple-pwas-on-the-same-domain/index.md:
```diff
@@ -0,0 +1,264 @@
+
+- **Storage:** Cookies, local storage, and all forms of device-local storage are
+  shared between apps. For that reason, if the user decides to wipe local data
+  for one app, it will wipe all the data from the origin; there's no way of
+  indicating to do this for a "single app". Note that Chrome and some other
```
```suggestion
  for one app, it will wipe all the data from the origin; there's no way 
 to do this for a single app. Note that Chrome and some other
```

jpmedley commented on src/site/content/en/blog/building-multiple-pwas-on-the-same-domain/index.md:
```diff
@@ -0,0 +1,264 @@
+
+## Challenges for multiple, same-origin PWAs
+
+Here are some practical issues common to both "same-origin" approaches:
```
```suggestion
Here are some practical issues common to both same-origin approaches:
```

jpmedley commented on src/site/content/en/blog/building-multiple-pwas-on-the-same-domain/index.md:
```diff
@@ -0,0 +1,264 @@
+  indicating to do this for a "single app". Note that Chrome and some other
+  browsers will actively prompt users to wipe local data when uninstalling one
+  of the apps, and this will affect data for the other apps on the origin as
+  well!
```
```suggestion
  well.
```

jpmedley commented on src/site/content/en/blog/building-multiple-pwas-on-the-same-domain/index.md:
```diff
@@ -0,0 +1,264 @@
+  simultaneously. That may sound like a good thing (not having to ask for a
+  permission multiple times), but remember: if the user blocks permission to one
+  app, it will prevent the others from requesting that permission or using that
+  feature!
```
```suggestion
  feature.
```

jpmedley commented on src/site/content/en/blog/building-multiple-pwas-on-the-same-domain/index.md:
```diff
@@ -0,0 +1,264 @@
+  setting to the other apps as well.
+
+These challenges make it difficult to encourage this approach. Nevertheless, if
+you can't use a separate origin (e.g. a subdomain), as discussed in the "Using
+separate origins" section, from the two "same-origin" options we presented,
```
```suggestion
you can't use a separate origin (e.g. a subdomain), as discussed in the [Using
separate origins](#using-separate-origins) section, from the two same-origin options we presented,
```

jpmedley commented on src/site/content/en/blog/building-multiple-pwas-on-the-same-domain/index.md:
```diff
@@ -0,0 +1,264 @@
+These challenges make it difficult to encourage this approach. Nevertheless, if
+you can't use a separate origin (e.g. a subdomain), as discussed in the "Using
+separate origins" section, from the two "same-origin" options we presented,
+using **non-overlapping paths** is strongly recommended, over
+**overlapping/nested paths**.
```
```suggestion
using non-overlapping paths is strongly recommended, over
overlapping/nested paths.
```

jpmedley commented on src/site/content/en/blog/building-multiple-pwas-on-the-same-domain/index.md:
```diff
@@ -0,0 +1,264 @@
+**overlapping/nested paths**.
+
+As mentioned, the challenges discussed in this section are common to both
+"same-origin" approaches. In the next section we'll go deeper into the details
```
```suggestion
same-origin approaches. In the next section we'll go deeper into the details
```

jpmedley commented on src/site/content/en/blog/building-multiple-pwas-on-the-same-domain/index.md:
```diff
@@ -0,0 +1,264 @@
+
+As mentioned, the challenges discussed in this section are common to both
+"same-origin" approaches. In the next section we'll go deeper into the details
+of why using **overlapping/nested paths** is the least recommended strategy.
```
```suggestion
of why using overlapping/nested paths is the least recommended strategy.
```

jpmedley commented on src/site/content/en/blog/building-multiple-pwas-on-the-same-domain/index.md:
```diff
@@ -0,0 +1,264 @@
+## Additional challenges for overlapping/nested paths
+
+The additional issue with overlapping/nested paths approach (where
+`https://example.com/` is the "outer app" and  `https://example.com/app/` is the
```
```suggestion
`https://example.com/` is the outer app and  `https://example.com/app/` is the
```

jpmedley commented on src/site/content/en/blog/building-multiple-pwas-on-the-same-domain/index.md:
```diff
@@ -0,0 +1,264 @@
+
+The additional issue with overlapping/nested paths approach (where
+`https://example.com/` is the "outer app" and  `https://example.com/app/` is the
+"inner app"), is that all URLs in the "inner app" will actually be considered to
```
```suggestion
inner app), is that all URLs in the inner app will actually be considered
```

jpmedley commented on src/site/content/en/blog/building-multiple-pwas-on-the-same-domain/index.md:
```diff
@@ -0,0 +1,264 @@
+The additional issue with overlapping/nested paths approach (where
+`https://example.com/` is the "outer app" and  `https://example.com/app/` is the
+"inner app"), is that all URLs in the "inner app" will actually be considered to
+be part of both the "outer app" and the "inner app".
```
```suggestion
part of both the outer app and the inner app.
```

jpmedley commented on src/site/content/en/blog/building-multiple-pwas-on-the-same-domain/index.md:
```diff
@@ -0,0 +1,264 @@
+
+In practice this presents the following issues:
+
+- **Installation Promotion:** If the user visits the "inner app" (for example,
```
```suggestion
- **Installation Promotion:** If the user visits the inner app (for example,
```

jpmedley commented on src/site/content/en/blog/building-multiple-pwas-on-the-same-domain/index.md:
```diff
@@ -0,0 +1,264 @@
+In practice this presents the following issues:
+
+- **Installation Promotion:** If the user visits the "inner app" (for example,
+  in a web browser), when the "outer app" is already installed in the user's
```
```suggestion
  in a web browser), when the outer app is already installed in the user's
```

jpmedley commented on src/site/content/en/blog/building-multiple-pwas-on-the-same-domain/index.md:
```diff
@@ -0,0 +1,264 @@
+  [BeforeInstallPrompt event](https://web.dev/customize-install/) won't be
+  triggered. The reason is that the browser will check and see whether the
+  current page belongs to an app that's already installed, and it will conclude
+  that it is. The workarounds for this is to install the "inner app" manually
```
```suggestion
  that it is. The workaround for this is to install the inner app manually
```

jpmedley commented on src/site/content/en/blog/building-multiple-pwas-on-the-same-domain/index.md:
```diff
@@ -0,0 +1,264 @@
+  triggered. The reason is that the browser will check and see whether the
+  current page belongs to an app that's already installed, and it will conclude
+  that it is. The workarounds for this is to install the "inner app" manually
+  (via "Create Shortcut" browser menu option), or to install the "inner app"
```
```suggestion
  (via "Create Shortcut" browser menu option), or to install the inner app
```

jpmedley commented on src/site/content/en/blog/building-multiple-pwas-on-the-same-domain/index.md:
```diff
@@ -0,0 +1,264 @@
+  current page belongs to an app that's already installed, and it will conclude
+  that it is. The workarounds for this is to install the "inner app" manually
+  (via "Create Shortcut" browser menu option), or to install the "inner app"
+  first before the "outer app".
```
```suggestion
  first, before the outer app.
```

jpmedley commented on src/site/content/en/blog/building-multiple-pwas-on-the-same-domain/index.md:
```diff
@@ -0,0 +1,264 @@
+  (via "Create Shortcut" browser menu option), or to install the "inner app"
+  first before the "outer app".
+- **[Notification](https://developer.mozilla.org/en-US/docs/Web/API/notification)
+  and [Badging API](https://web.dev/badging-api/)**: If the "outer app" is
```
```suggestion
  and the [Badging API](https://web.dev/badging-api/)**: If the outer app is
```

jpmedley commented on src/site/content/en/blog/building-multiple-pwas-on-the-same-domain/index.md:
```diff
@@ -0,0 +1,264 @@
+  first before the "outer app".
+- **[Notification](https://developer.mozilla.org/en-US/docs/Web/API/notification)
+  and [Badging API](https://web.dev/badging-api/)**: If the "outer app" is
+  installed but the "inner app" is not, notifications and badges coming from the
```
```suggestion
  installed but the inner app is not, notifications and badges coming from the
```

jpmedley commented on src/site/content/en/blog/building-multiple-pwas-on-the-same-domain/index.md:
```diff
@@ -0,0 +1,264 @@
+- **[Notification](https://developer.mozilla.org/en-US/docs/Web/API/notification)
+  and [Badging API](https://web.dev/badging-api/)**: If the "outer app" is
+  installed but the "inner app" is not, notifications and badges coming from the
+  "inner app" will be erroneously attributed to the "outer app" (which is the
```
```suggestion
  inner app will be erroneously attributed to the outer app (which is the
```

jpmedley commented on src/site/content/en/blog/building-multiple-pwas-on-the-same-domain/index.md:
```diff
@@ -0,0 +1,264 @@
+  the case where both apps are installed on the user's device.
+- **[Link
+  Capturing](https://github.com/WICG/sw-launch/blob/main/declarative_link_capturing.md)**:
+  The "outer app" may capture URLs that belong to the "inner app". This is
```
```suggestion
  The outer app may capture URLs that belong to the inner app. This is
```

jpmedley commented on src/site/content/en/blog/building-multiple-pwas-on-the-same-domain/index.md:
```diff
@@ -0,0 +1,264 @@
+- **[Link
+  Capturing](https://github.com/WICG/sw-launch/blob/main/declarative_link_capturing.md)**:
+  The "outer app" may capture URLs that belong to the "inner app". This is
+  especially likely if the "outer app" is installed but the "inner app" isn't.
```
```suggestion
  especially likely if the outer app is installed but the inner app isn't.
```

jpmedley commented on src/site/content/en/blog/building-multiple-pwas-on-the-same-domain/index.md:
```diff
@@ -0,0 +1,264 @@
+  Capturing](https://github.com/WICG/sw-launch/blob/main/declarative_link_capturing.md)**:
+  The "outer app" may capture URLs that belong to the "inner app". This is
+  especially likely if the "outer app" is installed but the "inner app" isn't.
+  Similarly, links within the "outer app" that link to the "inner app" will not
```
```suggestion
  Similarly, links within the outer app that link to the inner app will not
```

jpmedley commented on src/site/content/en/blog/building-multiple-pwas-on-the-same-domain/index.md:
```diff
@@ -0,0 +1,264 @@
+  The "outer app" may capture URLs that belong to the "inner app". This is
+  especially likely if the "outer app" is installed but the "inner app" isn't.
+  Similarly, links within the "outer app" that link to the "inner app" will not
+  link capture into the "inner app", since they are considered to be within the
```
```suggestion
  link capture into the inner app, since they are considered to be within the
```

jpmedley commented on src/site/content/en/blog/building-multiple-pwas-on-the-same-domain/index.md:
```diff
@@ -0,0 +1,264 @@
+  especially likely if the "outer app" is installed but the "inner app" isn't.
+  Similarly, links within the "outer app" that link to the "inner app" will not
+  link capture into the "inner app", since they are considered to be within the
+  "outer app"'s scope. Additionally, on Chrome OS / Android, if these apps are
```
```suggestion
  outer app's scope. Additionally, on Chrome OS and Android, if these apps are
```

jpmedley commented on src/site/content/en/blog/building-multiple-pwas-on-the-same-domain/index.md:
```diff
@@ -0,0 +1,264 @@
+  "outer app"'s scope. Additionally, on Chrome OS / Android, if these apps are
+  added to the Play Store (as [Trusted Web
+  Activities](https://developer.chrome.com/docs/android/trusted-web-activity/overview/)),
+  the "outer app" will capture all links. Even if the "inner app" is installed,
```
```suggestion
  the outer app will capture all links. Even if the inner app is installed,
```

jpmedley commented on src/site/content/en/blog/building-multiple-pwas-on-the-same-domain/index.md:
```diff
@@ -0,0 +1,264 @@
+  added to the Play Store (as [Trusted Web
+  Activities](https://developer.chrome.com/docs/android/trusted-web-activity/overview/)),
+  the "outer app" will capture all links. Even if the "inner app" is installed,
+  the OS will still offer the user the choice of opening them in the "outer
+  app."
```
```suggestion
  the OS will still offer the user the choice of opening them in the outer
  app.
```

jpmedley commented on src/site/content/en/blog/building-multiple-pwas-on-the-same-domain/index.md:
```diff
@@ -0,0 +1,264 @@
+## Conclusion
+
+In this article we looked at different ways in which developers can build
+multiple Progressive Web Apps related to each other within the same **domain
+name**.
```
```suggestion
multiple Progressive Web Apps related to each other within the same **domain**.
```

jpmedley replied:
This is to be consistent with what you've done elsewhere.

jpmedley commented on src/site/content/en/blog/building-multiple-pwas-on-the-same-domain/index.md:
```diff
@@ -0,0 +1,264 @@
+
+In summary, we strongly recommend using a different origin (e.g. by using
+subdomains) to host independent PWAs. Hosting them in the same origin presents
+many challenges, mainly due to the fact that the browser won't fully consider
```
```suggestion
many challenges, mainly because the browser won't fully consider
```

jpmedley commented on src/site/content/en/blog/building-multiple-pwas-on-the-same-domain/index.md:
```diff
@@ -0,0 +1,264 @@
+In summary, we strongly recommend using a different origin (e.g. by using
+subdomains) to host independent PWAs. Hosting them in the same origin presents
+many challenges, mainly due to the fact that the browser won't fully consider
+these to be distinct "apps".
```
```suggestion
these to be distinct apps.
```

jpmedley commented on src/site/content/en/blog/building-multiple-pwas-on-the-same-domain/index.md:
```diff
@@ -0,0 +1,264 @@
+If it's not possible to use different origins, using non-overlapping paths (e.g.
+`https://example.com/app1/` and `https://example.com/app2/` it's strongly
+recommended over using overlapping or nested paths, like `https://example.com/`
+(for the "outer app") and `https://example.com/app/` (for the "inner app").
```
```suggestion
(for the outer app) and `https://example.com/app/` (for the inner app).
```

jpmedley reviewed (changes requested):
Most of my requests are related to web.dev usage standards, though there are a few wording suggestions.

demianrenzulli commented on src/site/content/en/blog/building-multiple-pwas-on-the-same-domain/index.md:
```diff
@@ -0,0 +1,264 @@
+  to work as individual apps, tied to the company's name.
+
+<figure class="w-figure">
+  {% Img src="image/26V1DWN36MZr3mUo8ChSBlCpzp43/XEwhGkkV9Bz7Z2KEJWjm.png", alt="Avoid using different origins for site sections of the same site when trying to build a unified Progresive Web App.", width="800", height="475" %}
```
Hi @jpmedley! That's what I wanted to say. There are site that have the home page at https://www.example.com (origin 1) and the checkout page at https://checkout.example.com (origin 2), and the login page at https://login.example.com (origin 3). This is actually a unique site, that, for some reason has been divided into multiple origins by using subdomains. If the developer wants to provide a single PWA experience across the site, that will have many challenges. That's why that alternative is not recommended.

Please, let me know if that's clear and if there's a better way of explaining it in the alternative image text.

demianrenzulli commented on src/site/content/en/blog/building-multiple-pwas-on-the-same-domain/index.md:
```diff
@@ -0,0 +1,264 @@
+using path-level scoping. However, in both cases, using the same origin presents
+many problems and limitations, the root of which stems from the fact that the
+browser won't fully consider these to be distinct "apps", therefore **this
+approach is discouraged**.
```
I understand that in most cases we tend to talk about the "positive" case and avoid talking too much about the "negative" one and just say that it's not recommended. 
In this particular article, talking about the non-recommended use case is important because is a very common mistake developers do, and they don't really understand why it's a mistake.

Let me know if that makes sense.

cc // @mgiuca

demianrenzulli commented:
Thanks for approving these changes, @jpmedley! We are still waiting for the profile details of one co-author, but if we don't receive them by today, I guess we can move forward and then add him at a later point.

I'll let you know when we are ready.

demianrenzulli commented:
Hi @jpmedley! We just added the latest author, so we should be good to go now.

Two questions:

- I think we talked about this for other sites, but I'm not totally aware of the rules for Web.dev: should I merge directly?
- The article has an old date (date: 2021-05-18 attribute). I assume I should update this to the tentative date of publication? It seems to me that it would be good to get this published after the long weekend (Wednesday, June 2nd sounds like a good date).

Let me know what you think!

jpmedley commented:
Do you have a squash and merge button that's not grayed out? If you do, you don't need my permission to use it.

To make it show up at a later date, put whatever date you want on it and add `scheduled: true` to the metadata. I think you'll need another approval from someone in DevRel, but I'm not sure about that. I've cc'd our contract tech writer since I'm about to be out of the office for several days.

cc: @rachelandrew

demianrenzulli commented:
Thanks for your reply @jpmedley!

I can confirm I see the button in green, so I should be able to do the merge myself.
I just changed the publication date for next Wednesday, June 2nd and I added `scheduled: true` in the metadata.

@rachelandrew: please, let me know if you need to approve this as well, or if it's ok for me to proceed and merge.

Thanks!

jpmedley commented:
I can confirm that she does not. We didn't mean to give you that impression. She's in a lot of discussions where her approval isn't required because as a contractor, her ability to self-select work is a little more limited than ours. We're just trying to keep her plate full.

jpmedley commented:
We do have another problem though. "cmp" needs to be added to the authors file before this can merge. Do you need help with that?

demianrenzulli commented:
Thanks @jpmedley!

Based on your response, I was about to merge this myself, but now I found that the linter is complaining that the recently created author "cmp" is missing.

@robdodson: my apologies if you are not the one to help on these cases. This article contains two new authors ("mgiuca" and "cmp"). As you can see both have been added to the `authorsData.json` file in this PR. 

I ran the linter before the latest commits I made and also checked out the PR locally once again and run the linter there and it works fine.

I'm wondering if the deployed version of the linter detects issues that the local version doesn't? 
Two ideas:

- The only special thing with the "cmp" author is that we haven't provided a profile image, since the author doesn't want to use one at this point. This field doesn't seem to be mandatory, but it might be causing some issues?
- The latest commit I did was changing the publication date for June 2nd and adding the `scheduled: true` attribute. Things seem to have failed after that.

We are very close to wrap this up, so thanks in advance for all the help.

demianrenzulli commented:
@jpmedley @robdodson I can confirm that after running `npm run test` locally in the PR I was able to find the issue:

The tests complain about the missing author, but that works fine if the attribute is not present. It seems like the tests break when adding the `scheduled: true` attribute in the metadata (which was [my latest commit](https://github.com/GoogleChrome/web.dev/pull/5389/commits/b48244531973d29248f659d23af60abe4150db55)).

I'm wondering if this is a bug in web.dev? The use case seems to be adding a new author at the same time of creating an article with a late publication date and the `scheduled: true` attribute.

robdodson commented:
@demianrenzulli I think I see what's going on. Assigning to @MichaelSolati to take a look. We should be able to get this fixed before the post's publish date.

demianrenzulli commented:
@robdodson thanks for the prompt reply and let me know if you need me to test anything else on my side 👍

MichaelSolati commented:
@demianrenzulli i think I fixed this just now, please try rebasing main on top of your branch and the tests should pass.

robdodson commented:
@demianrenzulli I think you're correct, I wasn't able to get the tests to pass with that PR either. Apologies for the confusion! We can get it fixed up on Tuesday after folks are back from the holidays. I think that still leaves one more day before the due date.

demianrenzulli commented:
Thanks for confirming @robdodson! By the way, Wednesday was a tentative (not a hard) deadline, so if it takes longer to get this fixed, we can always publish the following week.

MichaelSolati commented:
Ok I think I misunderstood the issue and solved something completely irrelevant. I just opened #5484 which should address your issue, I believe it should... @robdodson will take a look at it on Tuesday I guess and then we can land it probably first thing on Tuesday as well.

demianrenzulli commented:
Hi folks,

I assume that once the fix is done you would be able to merge this PR directly? The publishing date is supposed to be tomorrow, but just let me know if you need more time to fix it, so we can push it a few more days.

Cheers.

MichaelSolati commented:
Not yet, just testing it, but should be ready soon we think.

MichaelSolati commented:
@demianrenzulli ok, I believe that the `main` branch should now have the fix for your issue. I tested it locally. @robdodson I'm unsure if this will introduce any new issues, but I think we should be fine.

demianrenzulli commented:
Thanks for the update @MichaelSolati!

This PR is blocked by the linter. If the changes are now in `main`, do you think that re-running the linter would help unblocking it? I'm not sure how often changes made in `main` end up hitting the servers where the linter runs.

robdodson commented:
I think this branch would need to be rebased to fix the failing tests. @MichaelSolati are you able to do that?

MichaelSolati commented:
Yes I can help, but can't at this second... I'm heading to the gym. @demianrenzulli do you think you'd be available in about 2 hours (4 PM PST)?

demianrenzulli commented:
@MichaelSolati yes! just ping me later and I'll be able to test it and merge this PR.

demianrenzulli commented:
Thanks for all the help @MichaelSolati! We tested the changes and they work well. Going to merge now. cc // @robdodson
